[Syntax Compatibility] CAST functions are not supported

Unable to parse [CAST](https://clickhouse.com/docs/en/sql-reference/functions/type-conversion-functions#castx-t) function.

Example SQL:
```
panic: line 0:24 expected keyword: AS, but got )
select cast(1 as Float64) as value
                        ^
```

Test Cases:

1. `select cast(1 as Float64) as value` 🚫 Not Working
2. `select cast(1, 'Float64') as value` 🚫 Not Working 
3. `select (1 as Float64) as value` ✅ Working
4. `select 1::Float64 as value` ✅ Working

--- a/parser/ast.go
+++ b/parser/ast.go
@@ -3430,12 +3430,9 @@ func (c *CastExpr) End() Pos {
 func (c *CastExpr) String(level int) string {
 	var builder strings.Builder
 	builder.WriteString("CAST(")
-	builder.WriteString(NewLine(level + 1))
 	builder.WriteString(c.Expr.String(level))
 	builder.WriteString(" AS ")
-	builder.WriteString(NewLine(level + 1))
 	builder.WriteString(c.AsType.String(level))
-	builder.WriteString(NewLine(level + 1))
 	builder.WriteByte(')')
 	return builder.String()
 }

--- a/parser/ast.go
+++ b/parser/ast.go
@@ -105,14 +105,18 @@ func (p *BinaryExpr) End() Pos {
 func (p *BinaryExpr) String(level int) string {
 	var builder strings.Builder
 	builder.WriteString(p.LeftExpr.String(level))
-	builder.WriteByte(' ')
+	if p.Operation != opTypeCast {
+		builder.WriteByte(' ')
+	}
 	if p.HasNot {
 		builder.WriteString("NOT ")
 	} else if p.HasGlobal {
 		builder.WriteString("GLOBAL ")
 	}
 	builder.WriteString(string(p.Operation))
-	builder.WriteByte(' ')
+	if p.Operation != opTypeCast {
+		builder.WriteByte(' ')
+	}
 	builder.WriteString(p.RightExpr.String(level))
 	return builder.String()
 }
@@ -3413,10 +3417,11 @@ func (c *CaseExpr) Accept(visitor ASTVisitor) error {
 }
 
 type CastExpr struct {
-	CastPos Pos
-	Expr    Expr
-	AsPos   Pos
-	AsType  Expr
+	CastPos   Pos
+	Expr      Expr
+	Separator string
+	AsPos     Pos
+	AsType    Expr
 }
 
 func (c *CastExpr) Pos() Pos {
@@ -3431,7 +3436,11 @@ func (c *CastExpr) String(level int) string {
 	var builder strings.Builder
 	builder.WriteString("CAST(")
 	builder.WriteString(c.Expr.String(level))
-	builder.WriteString(" AS ")
+	if c.Separator == "," {
+		builder.WriteString(", ")
+	} else {
+		builder.WriteString(" AS ")
+	}
 	builder.WriteString(c.AsType.String(level))
 	builder.WriteByte(')')
 	return builder.String()

--- a/parser/lexer.go
+++ b/parser/lexer.go
@@ -15,8 +15,6 @@ const (
 	TokenInt     TokenKind = "<int>"
 	TokenFloat   TokenKind = "<float>"
 	TokenString  TokenKind = "<string>"
-	TokenCast    TokenKind = "<cast>"
-	TokenArrow   TokenKind = "<arrow>"
 )
 
 type Pos int
@@ -276,7 +274,7 @@ func (l *Lexer) consumeToken() error {
 		} else if l.peekOk(1) && l.peekN(1) == '>' {
 			l.lastToken = &Token{
 				String: l.slice(0, 2),
-				Kind:   TokenArrow,
+				Kind:   opTypeArrow,
 				Pos:    Pos(l.current),
 				End:    Pos(l.current + 2),
 			}
@@ -295,7 +293,7 @@ func (l *Lexer) consumeToken() error {
 		if l.peekOk(1) && l.peekN(1) == ':' {
 			l.lastToken = &Token{
 				String: l.slice(0, 2),
-				Kind:   TokenCast,
+				Kind:   opTypeCast,
 				Pos:    Pos(l.current),
 				End:    Pos(l.current + 2),
 			}

--- a/parser/parser_column.go
+++ b/parser/parser_column.go
@@ -388,14 +388,18 @@ func (p *Parser) parseColumnCastExpr(pos Pos) (Expr, error) {
 		return nil, err
 	}
 
-	columnExpr, err := p.parseExpr(p.Pos())
+	columnExpr, err := p.parseColumnExpr(p.Pos())
 	if err != nil {
 		return nil, err
 	}
 
 	asPos := p.Pos()
-	if err := p.consumeKeyword(KeywordAs); err != nil {
-		return nil, err
+	switch {
+	// CAST(x, T) and CAST(x AS T) are equivalent
+	case p.matchKeyword(KeywordAs), p.matchTokenKind(","):
+		_ = p.lexer.consumeToken()
+	default:
+		return nil, fmt.Errorf("expected AS or , but got %s", p.lastTokenKind())
 	}
 	asColumnType, err := p.parseColumnType(p.Pos())
 	if err != nil {
@@ -645,6 +649,9 @@ func (p *Parser) parseColumnCaseExpr(pos Pos) (*CaseExpr, error) {
 }
 
 func (p *Parser) parseColumnType(_ Pos) (Expr, error) { // nolint:funlen
+	if p.matchTokenKind(TokenString) {
+		return p.parseString(p.Pos())
+	}
 	ident, err := p.parseIdent()
 	if err != nil {
 		return nil, err

--- a/parser/parser_column.go
+++ b/parser/parser_column.go
@@ -238,8 +238,8 @@ func (p *Parser) parseMulDivModExpr(pos Pos) (Expr, error) {
 		case p.matchTokenKind(opTypeMul),
 			p.matchTokenKind(opTypeDiv),
 			p.matchTokenKind(opTypeMod),
-			p.matchTokenKind(TokenArrow),
-			p.matchTokenKind(TokenCast):
+			p.matchTokenKind(opTypeArrow),
+			p.matchTokenKind(opTypeCast):
 			op := p.lastTokenKind()
 			_ = p.lexer.consumeToken()
 			rightExpr, err := p.parseUnaryExpr(p.Pos())
@@ -393,10 +393,12 @@ func (p *Parser) parseColumnCastExpr(pos Pos) (Expr, error) {
 		return nil, err
 	}
 
+	var separator string
 	asPos := p.Pos()
 	switch {
 	// CAST(x, T) and CAST(x AS T) are equivalent
 	case p.matchKeyword(KeywordAs), p.matchTokenKind(","):
+		separator = p.last().String
 		_ = p.lexer.consumeToken()
 	default:
 		return nil, fmt.Errorf("expected AS or , but got %s", p.lastTokenKind())
@@ -411,10 +413,11 @@ func (p *Parser) parseColumnCastExpr(pos Pos) (Expr, error) {
 	}
 
 	return &CastExpr{
-		CastPos: pos,
-		AsPos:   asPos,
-		Expr:    columnExpr,
-		AsType:  asColumnType,
+		CastPos:   pos,
+		AsPos:     asPos,
+		Separator: separator,
+		Expr:      columnExpr,
+		AsType:    asColumnType,
 	}, nil
 }
 

--- a/parser/parser_table.go
+++ b/parser/parser_table.go
@@ -1303,7 +1303,7 @@ func (p *Parser) parseCreateFunction(pos Pos) (*CreateFunction, error) {
 	if err != nil {
 		return nil, err
 	}
-	if _, err := p.consumeTokenKind(TokenArrow); err != nil {
+	if _, err := p.consumeTokenKind(opTypeArrow); err != nil {
 		return nil, err
 	}
 	expr, err := p.parseExpr(p.Pos())

--- a/parser/testdata/query/format/select_cast.sql
+++ b/parser/testdata/query/format/select_cast.sql
@@ -10,10 +10,10 @@ SELECT
   CAST(1 AS Float64) AS value;
 
 SELECT 
-  CAST(1 AS 'Float64') AS value;
+  CAST(1, 'Float64') AS value;
 
 SELECT 
   (1 AS Float64) AS value;
 
 SELECT 
-  1 <cast> Float64 AS value;
+  1::Float64 AS value;

--- a/parser/testdata/query/format/select_cast.sql
+++ b/parser/testdata/query/format/select_cast.sql
@@ -1,0 +1,19 @@
+-- Origin SQL:
+select cast(1 as Float64) as value;
+select cast(1, 'Float64') as value;
+select (1 as Float64) as value;
+select 1::Float64 as value;
+
+-- Format SQL:
+
+SELECT 
+  CAST(1 AS Float64) AS value;
+
+SELECT 
+  CAST(1 AS 'Float64') AS value;
+
+SELECT 
+  (1 AS Float64) AS value;
+
+SELECT 
+  1 <cast> Float64 AS value;

--- a/parser/testdata/query/output/select_cast.sql.golden.json
+++ b/parser/testdata/query/output/select_cast.sql.golden.json
@@ -1,0 +1,227 @@
+[
+  {
+    "SelectPos": 0,
+    "StatementEnd": 34,
+    "With": null,
+    "Top": null,
+    "SelectColumns": {
+      "ListPos": 7,
+      "ListEnd": 34,
+      "HasDistinct": false,
+      "Items": [
+        {
+          "Expr": {
+            "CastPos": 7,
+            "Expr": {
+              "NumPos": 12,
+              "NumEnd": 13,
+              "Literal": "1",
+              "Base": 10
+            },
+            "AsPos": 14,
+            "AsType": {
+              "Name": {
+                "Name": "Float64",
+                "Unquoted": false,
+                "NamePos": 17,
+                "NameEnd": 24
+              }
+            }
+          },
+          "AliasPos": 26,
+          "Alias": {
+            "Name": "value",
+            "Unquoted": false,
+            "NamePos": 29,
+            "NameEnd": 34
+          }
+        }
+      ]
+    },
+    "From": null,
+    "ArrayJoin": null,
+    "Window": null,
+    "Prewhere": null,
+    "Where": null,
+    "GroupBy": null,
+    "WithTotal": false,
+    "Having": null,
+    "OrderBy": null,
+    "LimitBy": null,
+    "Limit": null,
+    "Settings": null,
+    "UnionAll": null,
+    "UnionDistinct": null,
+    "Except": null
+  },
+  {
+    "SelectPos": 36,
+    "StatementEnd": 70,
+    "With": null,
+    "Top": null,
+    "SelectColumns": {
+      "ListPos": 43,
+      "ListEnd": 70,
+      "HasDistinct": false,
+      "Items": [
+        {
+          "Expr": {
+            "CastPos": 43,
+            "Expr": {
+              "NumPos": 48,
+              "NumEnd": 49,
+              "Literal": "1",
+              "Base": 10
+            },
+            "AsPos": 49,
+            "AsType": {
+              "LiteralPos": 52,
+              "LiteralEnd": 59,
+              "Literal": "Float64"
+            }
+          },
+          "AliasPos": 62,
+          "Alias": {
+            "Name": "value",
+            "Unquoted": false,
+            "NamePos": 65,
+            "NameEnd": 70
+          }
+        }
+      ]
+    },
+    "From": null,
+    "ArrayJoin": null,
+    "Window": null,
+    "Prewhere": null,
+    "Where": null,
+    "GroupBy": null,
+    "WithTotal": false,
+    "Having": null,
+    "OrderBy": null,
+    "LimitBy": null,
+    "Limit": null,
+    "Settings": null,
+    "UnionAll": null,
+    "UnionDistinct": null,
+    "Except": null
+  },
+  {
+    "SelectPos": 72,
+    "StatementEnd": 102,
+    "With": null,
+    "Top": null,
+    "SelectColumns": {
+      "ListPos": 79,
+      "ListEnd": 102,
+      "HasDistinct": false,
+      "Items": [
+        {
+          "Expr": {
+            "LeftParenPos": 79,
+            "RightParenPos": 92,
+            "Items": {
+              "ListPos": 80,
+              "ListEnd": 92,
+              "HasDistinct": false,
+              "Items": [
+                {
+                  "Expr": {
+                    "NumPos": 80,
+                    "NumEnd": 81,
+                    "Literal": "1",
+                    "Base": 10
+                  },
+                  "AliasPos": 82,
+                  "Alias": {
+                    "Name": "Float64",
+                    "Unquoted": false,
+                    "NamePos": 85,
+                    "NameEnd": 92
+                  }
+                }
+              ]
+            },
+            "ColumnArgList": null
+          },
+          "AliasPos": 94,
+          "Alias": {
+            "Name": "value",
+            "Unquoted": false,
+            "NamePos": 97,
+            "NameEnd": 102
+          }
+        }
+      ]
+    },
+    "From": null,
+    "ArrayJoin": null,
+    "Window": null,
+    "Prewhere": null,
+    "Where": null,
+    "GroupBy": null,
+    "WithTotal": false,
+    "Having": null,
+    "OrderBy": null,
+    "LimitBy": null,
+    "Limit": null,
+    "Settings": null,
+    "UnionAll": null,
+    "UnionDistinct": null,
+    "Except": null
+  },
+  {
+    "SelectPos": 104,
+    "StatementEnd": 130,
+    "With": null,
+    "Top": null,
+    "SelectColumns": {
+      "ListPos": 111,
+      "ListEnd": 130,
+      "HasDistinct": false,
+      "Items": [
+        {
+          "Expr": {
+            "LeftExpr": {
+              "NumPos": 111,
+              "NumEnd": 112,
+              "Literal": "1",
+              "Base": 10
+            },
+            "Operation": "\u003ccast\u003e",
+            "RightExpr": {
+              "Name": "Float64",
+              "Unquoted": false,
+              "NamePos": 114,
+              "NameEnd": 121
+            },
+            "HasGlobal": false,
+            "HasNot": false
+          },
+          "AliasPos": 122,
+          "Alias": {
+            "Name": "value",
+            "Unquoted": false,
+            "NamePos": 125,
+            "NameEnd": 130
+          }
+        }
+      ]
+    },
+    "From": null,
+    "ArrayJoin": null,
+    "Window": null,
+    "Prewhere": null,
+    "Where": null,
+    "GroupBy": null,
+    "WithTotal": false,
+    "Having": null,
+    "OrderBy": null,
+    "LimitBy": null,
+    "Limit": null,
+    "Settings": null,
+    "UnionAll": null,
+    "UnionDistinct": null,
+    "Except": null
+  }
+]

--- a/parser/testdata/query/output/select_cast.sql.golden.json
+++ b/parser/testdata/query/output/select_cast.sql.golden.json
@@ -18,6 +18,7 @@
               "Literal": "1",
               "Base": 10
             },
+            "Separator": "as",
             "AsPos": 14,
             "AsType": {
               "Name": {
@@ -73,6 +74,7 @@
               "Literal": "1",
               "Base": 10
             },
+            "Separator": ",",
             "AsPos": 49,
             "AsType": {
               "LiteralPos": 52,
@@ -188,7 +190,7 @@
               "Literal": "1",
               "Base": 10
             },
-            "Operation": "\u003ccast\u003e",
+            "Operation": "::",
             "RightExpr": {
               "Name": "Float64",
               "Unquoted": false,

--- a/parser/testdata/query/select_cast.sql
+++ b/parser/testdata/query/select_cast.sql
@@ -1,0 +1,4 @@
+select cast(1 as Float64) as value;
+select cast(1, 'Float64') as value;
+select (1 as Float64) as value;
+select 1::Float64 as value;

--- a/parser/type.go
+++ b/parser/type.go
@@ -22,6 +22,9 @@ const (
 	opTypeDiv   TokenKind = "/"
 	opTypeMod   TokenKind = "%"
 
+	opTypeArrow TokenKind = "->"
+	opTypeCast  TokenKind = "::"
+
 	// Logical operators
 	opTypeAnd TokenKind = "AND"
 	opTypeOr  TokenKind = "OR"

